torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -332,7 +332,8 @@ void check_gpu_tensors(const std::vector<at::Tensor>& tensors) {
 std::vector<at::Tensor> flatten_for_scatter_gather(
     std::vector<std::vector<at::Tensor>>& tensor_lists,
     std::vector<at::Tensor>& other,
-    size_t world_size) {
+    size_t world_size,
+    size_t rank) {
```

Comment:
to simplify reasoning in upstream code, can you introduce 
bool* inplace or bool* already_flat instead and fill it in as you do below for 'inplace'?

Reply:
Do you mean introduce a new flag in python API so that "discovery" logic can be removed from ProcessGroupNCCL.cpp? In other words

torch.distributed.all_gather(output_tensors, input_tensor)

Will always flatten output tensors, while

torch.distributed.all_gather(output_tensors, input_tensor, already_flat=True)

will always do an inplace reduction?

Reply:
This can be an option as well, but I think the point of this cl is to allow user code to run fast without necessary changing it explicitly, so they'll get it 'for free'.

I meant having output bool here, e.g.
auto outputFlattened = flatten_for_scatter_gather(
    outputTensors, inputTensors, size, rank, *output_is_flat
  );

and then use it instead of

 if (outputFlattened[i][j].storage().is_alias_of(
              outputTensors[i][j].storage()) &&
              outputTensors[i][j].storage_offset() ==
              (outputTensors[i][0].storage_offset() +
              outputTensors[i][j].numel() * j)) {
             break;
          }

Would it work?

Reply:
It will work, but the logic that determines if the list of tensors are views into an already flattened tensor will then have to go into the code that implements each NCCL op, so we'd end up duplicating it. Unless I put the reasoning code in a separate method.